harness/determined/exec/worker_process_wrapper.py

```diff
@@ -4,23 +4,48 @@
 to package a shell script.
 """
 import os
+import subprocess
 import sys
+import threading
+from typing import BinaryIO, List
 
 from determined import constants
 
 
+def forward_stream(src_stream: BinaryIO, dst_stream: BinaryIO, rank: str) -> None:
+    for line in iter(src_stream.readline, b""):
+        line = f"[rank={rank}] ".encode() + line
+        os.write(dst_stream.fileno(), line)
+
+
+def run_all(ts: List[threading.Thread]) -> None:
+    for t in ts:
+        t.start()
+    for t in ts:
+        t.join()
+
+
 def main() -> None:
-    os.dup2(os.open(constants.CONTAINER_STDOUT, os.O_WRONLY), sys.stdout.fileno())
-    os.dup2(os.open(constants.CONTAINER_STDERR, os.O_WRONLY), sys.stderr.fileno())
-    os.execv(
-        sys.executable,
+    rank = os.environ.get("HOROVOD_RANK")
+    proc = subprocess.Popen(
         [
             sys.executable,
             "-m",
             "determined.exec.worker_process",
             *sys.argv[1:],
         ],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
     )
+    with open(constants.CONTAINER_STDOUT, "w") as cstdout, open(
+        constants.CONTAINER_STDERR, "w"
+    ) as cstderr, proc:
```

Comment:
This is what the formatter wants this to be 🤷

Reply:
Heh, I had a feeling it might do something strange, which is why I mentioned "if that looks any better". There's no actual disadvantage to having three separate logical lines (the bytecode ends up identical), so feel free to put it back the other way.